### PR TITLE
fmt, info, lint, sync: extract common types

### LINKED ISSUE #558: fmt, info, lint, sync: use common types
Erik is working on this.

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Feel free to change the commit message whatever you like

Closes: #558

### CODE REVIEW

File: src/info/info.nim [hunk 1]
```diff
@@ -91,7 +91,7 @@ proc conceptsInfo(practiceExercises: seq[PracticeExercise],
 func getSlugs(practiceExercises: seq[PracticeExercise]): HashSet[string] =
   result = initHashSet[string](practiceExercises.len)
   for practiceExercise in practiceExercises:
-    result.incl practiceExercise.slug
+    result.incl $practiceExercise.slug
```
Comment (ErikSchierboom) [inline src/info/info.nim:93]: Previously, the slug was a `string`, but now it is a `Slug`

File: src/sync/sync_common.nim [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import std/[algorithm, enumutils, json, options, os, sets, strformat, strutils]
 import pkg/jsony
-import ".."/[cli, helpers, lint/validators]
+import ".."/[cli,helpers, lint/validators, types_exercise_config, types_track_config]
```
Comment (ee7) [inline src/sync/sync_common.nim]:
```suggestion
import ".."/[cli, helpers, lint/validators, types_exercise_config, types_track_config]
```

File: src/types_track_config.nim [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  Concept* = object
+    name*: string
+    slug*: string
+    uuid*: string
+
+  Concepts* = seq[Concept]
+
+  TrackConfig* = object
+    slug*: string
```
Comment (ee7) [inline src/types_track_config.nim:51]: This makes me ask: is there a reason that this PR changed to `slug*: Slug` above, but not here? (I'll have a look).

File: src/lint/track_config.nim [hunk 4]
```diff
@@ -1,8 +1,10 @@
 import std/[json, os, sets, strformat, strutils, tables]
 import pkg/jsony
-import ".."/[cli, helpers]
+import ".."/[cli, helpers, types_track_config]
 import "."/validators
 
+func `$`*(slug: Slug): string {.borrow.}
+
```
Comment (ee7) [inline src/lint/track_config.nim]: Hmm. This brings us to a sort of intermediate state where e.g. `info.nim` still has to import `lint/track_config`, but just to get `$` and the `TrackConfig.init`. I think we want to move the common procs that operate on a type into a common module too. I'll have a look.

File: src/types_track_config.nim (hunk 3 above)
Comment (ErikSchierboom) [reply src/types_track_config.nim:51]: Not really (I think)

File: src/lint/track_config.nim (hunk 4 above)
Comment (ErikSchierboom) [reply src/lint/track_config.nim]: Great. You'll have a better understanding of the code base than I do.

**Review by ee7 — COMMENTED** [review]
There's some subtlety in the slug change.

The idea was that we only initialize a `Slug` with a known kebab-case value. With the current state, `lint` now uses `Slug`s that may not be kebab case.

We could move the `postHook` from `sync_common` to `types_track_config`, make other things `Slug`, and then remove some kebab-case checking in `lint`. But that's technically a behavior change for `lint` - it would exit on the first non-kebab slug without reporting any other problems. And this PR is complex enough anyway, so let's leave everything else for later. We'll also try to make `fmt` not import `sync` stuff.

I'll merge soon, I just want to:
- Do a sanity check of the lint/sync output on every track
- Briefly consider reverting everything from the commit I added apart from the `$` changes.

Also for later: there's a pre-existing bug in ` tidyJsonyErrorMsg` - it shouldn't say "configlet lint", since it isn't used only by `lint`. See https://github.com/exercism/configlet/commit/6b215a3311344377eeb9f0aadb5f941b6e930b93 for background.

**Review by ee7 — APPROVED** [review]
This PR preserves the output of lint and sync on every track.

Resolved merge conflict due to https://github.com/exercism/configlet/commit/152a953676a9882cdc5df58a7ac46008e09930da.

Branch diff before force-push: https://github.com/exercism/configlet/compare/4.0.0-beta.1..e079de9834f80b8c01f5dbb0d23b7f5cde50fd0a

Branch diff after force-push: https://github.com/exercism/configlet/compare/4.0.0-beta.2..23cad1fee0e2890423d247ec6e9dd900991581f8

```console
$ git diff 4.0.0-beta.1 e079de98 | grep '^[+-]' > /tmp/before.txt
$ git diff 4.0.0-beta.2 23cad1fe | grep '^[+-]' > /tmp/after.txt
$ git diff /tmp/before.txt /tmp/after.txt > /tmp/diff_between_diffs.txt
```

```diff
diff --git a/tmp/before.txt b/tmp/after.txt
--- a/tmp/before.txt
+++ b/tmp/after.txt
@@ -84,6 +84,7 @@
 -    example*: seq[string]
 -    exemplar*: seq[string]
 -    editor*: seq[string]
+-    invalidator*: seq[string]
 -
 -  TrackConfig* = object
 -    slug*: string
@@ -179,6 +180,7 @@
 -    exemplar*: seq[string]
 -    example*: seq[string]
 -    editor*: seq[string]
+-    invalidator*: seq[string]
 -
 -  TrackConfig* = object
 -    exercises*: Exercises
@@ -229,6 +231,7 @@
 -  #     solution*: seq[string]
 -  #     test*: seq[string]
 -  #     editor*: seq[string]
+-  #     invalidator*: seq[string]
 -  #     case kind*: ExerciseKind
 -  #     of ekConcept:
 -  #       exemplar*: seq[string]
@@ -261,6 +264,7 @@
 -    fkExemplar = "exemplar"
 -    fkExample = "example"
 -    fkEditor = "editor"
+-    fkInvalidator = "invalidator"
 -
 -  ConceptExerciseFiles* = object
 -    originalKeyOrder: seq[FilesKey]
@@ -268,6 +272,7 @@
 -    test*: seq[string]
 -    exemplar*: seq[string]
 -    editor*: seq[string]
+-    invalidator*: seq[string]
 -
 -  PracticeExerciseFiles* = object
 -    originalKeyOrder: seq[FilesKey]
@@ -275,6 +280,7 @@
 -    test*: seq[string]
 -    example*: seq[string]
 -    editor*: seq[string]
+-    invalidator*: seq[string]
 -
 -  ConceptExerciseConfig* = object
 -    originalKeyOrder: seq[ExerciseConfigKey]
@@ -360,6 +366,7 @@
 +  #     solution*: seq[string]
 +  #     test*: seq[string]
 +  #     editor*: seq[string]
++  #     invalidator*: seq[string]
 +  #     case kind*: ExerciseKind
 +  #     of ekConcept:
 +  #       exemplar*: seq[string]
@@ -392,6 +399,7 @@
 +    fkExemplar = "exemplar"
 +    fkExample = "example"
 +    fkEditor = "editor"
++    fkInvalidator = "invalidator"
 +
 +  ConceptExerciseFiles* = object
 +    originalKeyOrder*: seq[FilesKey]
@@ -399,6 +407,7 @@
 +    test*: seq[string]
 +    exemplar*: seq[string]
 +    editor*: seq[string]
++    invalidator*: seq[string]
 +
 +  PracticeExerciseFiles* = object
 +    originalKeyOrder*: seq[FilesKey]
@@ -406,6 +415,7 @@
 +    test*: seq[string]
 +    example*: seq[string]
 +    editor*: seq[string]
++    invalidator*: seq[string]
 +
 +  ConceptExerciseConfig* = object
 +    originalKeyOrder*: seq[ExerciseConfigKey]
@@ -477,6 +487,7 @@
 +    exemplar*: seq[string]
 +    example*: seq[string]
 +    editor*: seq[string]
++    invalidator*: seq[string]
 +
 +  Concept* = object
 +    name*: string
```

File: src/lint/track_config.nim (hunk 4 above)
Comment (ee7) [reply src/lint/track_config.nim]: Resolved this part.

File: src/types_track_config.nim (hunk 3 above)
Comment (ee7) [reply src/types_track_config.nim:51]: The answer was yes: this PR moved to `slug: Slug` only when that exact thing was a `Slug` in one place already, which is fine by me.

Tracking the others in https://github.com/exercism/configlet/issues/594.

File: src/types_exercise_config.nim [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  #     invalidator*: seq[string]
+  #     case kind*: ExerciseKind
+  #     of ekConcept:
+  #       exemplar*: seq[string]
+  #     of ekPractice:
+  #       example*: seq[string]
+  #
+  # (again because the JSON `files` data does not contain a `kind` key) we can
+  # theoretically determine the `kind` by the presence of the `exemplar` or
+  # `example` keys.
```
Comment (ErikSchierboom) [inline src/types_exercise_config.nim:51]: Can you hook into jsony somehow?

**Review by ErikSchierboom — COMMENTED** [review]
Thanks for getting this merged!

**Pull request merged**